File: react-native-nitro-player/ios/core/TrackPlayerShuffle.swift
//
//  TrackPlayerShuffle.swift
//  NitroPlayer
//

import Foundation

extension TrackPlayerCore {

  func seedShuffleOrder(_ tracks: [TrackItem], firstId: String? = nil) {
    let rest = tracks.map(\.id).filter { $0 != firstId }.shuffled()
    shuffleOrder = (firstId.map { [$0] } ?? []) + rest
  }

  /// Identity when shuffle is off; otherwise reorders `tracks` by `shuffleOrder` — stale ids drop out, new ids join a random upcoming slot.
  func applyShuffleOrder(_ tracks: [TrackItem]) -> [TrackItem] {
    guard shuffleEnabled else { return tracks }
    // ponytail: duplicate ids collapse to one entry, already unsupported everywhere else
    let byId = Dictionary(tracks.map { ($0.id, $0) }, uniquingKeysWith: { _, last in last })
    shuffleOrder.removeAll { byId[$0] == nil }
    var known = Set(shuffleOrder)
    let lo = min(max(currentTrackIndex + 1, 0), shuffleOrder.count)
    for track in tracks {
      guard known.insert(track.id).inserted else { continue }
      shuffleOrder.insert(track.id, at: Int.random(in: lo...shuffleOrder.count))
    }
    return shuffleOrder.compactMap { byId[$0] }
  }

  private func applyShuffleState() {
    let source = currentPlaylistId.flatMap { playlistManager.getPlaylist(playlistId: $0)?.tracks } ?? currentTracks
    let anchorId = currentTracks[safe: currentTrackIndex]?.id
    if shuffleEnabled { seedShuffleOrder(source, firstId: anchorId) }
    currentTracks = applyShuffleOrder(source)
    currentTrackIndex = anchorId.flatMap { id in currentTracks.firstIndex { $0.id == id } } ?? -1
    if isCasting || player?.currentItem != nil { rebuildAVQueueFromCurrentPosition() }
    checkUpcomingTracksForUrls(lookahead: lookaheadCount)
    notifyShuffleChange(shuffleEnabled)
  }

  func setShuffleModeOnQueue(enabled: Bool) {
    guard enabled != shuffleEnabled else { return }
    shuffleEnabled = enabled
    applyShuffleState()
  }

  func setShuffleMode(enabled: Bool) async {
    await withPlayerQueueNoThrow { self.setShuffleModeOnQueue(enabled: enabled) }
  }

  func reshuffleOnQueue() {
    shuffleEnabled = true
    applyShuffleState()
  }
}

File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/core/TrackPlayerShuffle.kt
@file:Suppress("ktlint:standard:max-line-length")

package com.margelo.nitro.nitroplayer.core

import com.margelo.nitro.nitroplayer.TrackItem
import kotlin.random.Random

internal fun TrackPlayerCore.seedShuffleOrder(
    tracks: List<TrackItem>,
    firstId: String? = null,
) {
    val rest = tracks.map { it.id }.filter { it != firstId }.shuffled()
    shuffleOrder = (listOfNotNull(firstId) + rest).toMutableList()
}

/** Identity when shuffle is off; otherwise reorders [tracks] by [shuffleOrder] — stale ids drop out, new ids join a random upcoming slot. */
internal fun TrackPlayerCore.applyShuffleOrder(tracks: List<TrackItem>): List<TrackItem> {
    if (!shuffleEnabled) return tracks
    val byId = tracks.associateBy { it.id } // ponytail: duplicate ids collapse to one entry, already unsupported everywhere else
    shuffleOrder.retainAll(byId.keys)
    val known = shuffleOrder.toHashSet()
    val lo = (currentTrackIndex + 1).coerceIn(0, shuffleOrder.size)
    for (track in tracks) {
        if (known.add(track.id)) shuffleOrder.add(Random.nextInt(lo, shuffleOrder.size + 1), track.id)
    }
    return shuffleOrder.mapNotNull { byId[it] }
}

private fun TrackPlayerCore.applyShuffleState() {
    val source = currentPlaylistId?.let { playlistManager.getPlaylist(it)?.tracks } ?: currentTracks
    val anchorId = currentTracks.getOrNull(currentTrackIndex)?.id
    if (shuffleEnabled) seedShuffleOrder(source, anchorId)
    currentTracks = applyShuffleOrder(source)
    currentTrackIndex = anchorId?.let { id -> currentTracks.indexOfFirst { it.id == id } } ?: -1
    if (isExoInitialized && exo.currentMediaItem != null) rebuildQueueFromCurrentPosition()
    checkUpcomingTracksForUrls(lookaheadCount)
    // Mirror into the local ExoPlayer only (CastPlayer ignores it); the order itself is ours, see ExoPlayerBuilder.
    val local = castSessionController?.localPlayer ?: if (isExoInitialized) exo.player else null
    if (local != null && local.shuffleModeEnabled != shuffleEnabled) local.shuffleModeEnabled = shuffleEnabled
    notifyShuffleChange(shuffleEnabled)
}

internal fun TrackPlayerCore.setShuffleModeOnQueue(enabled: Boolean) {
    if (enabled == shuffleEnabled) return
    shuffleEnabled = enabled
    applyShuffleState()
}

internal fun TrackPlayerCore.reshuffleOnQueue() {
    shuffleEnabled = true
    applyShuffleState()
}

fun TrackPlayerCore.getShuffleMode(): Boolean = shuffleEnabled
